src/auth/oauth2client.ts

```diff
@@ -660,20 +679,34 @@ export class OAuth2Client extends AuthClient {
     options: GetTokenOptions
   ): Promise<GetTokenResponse> {
     const url = this.endpoints.oauth2TokenUrl.toString();
+    const headers: Headers = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    };
     const values = {
       code: options.code,
       client_id: options.client_id || this._clientId,
-      client_secret: this._clientSecret,
       redirect_uri: options.redirect_uri || this.redirectUri,
       grant_type: 'authorization_code',
       code_verifier: options.codeVerifier,
     };
+    if (this.client_authentication === ClientAuthentication.ClientSecretBasic) {
+      const basic_auth =
+        'Basic ' +
+        Buffer.from(`${this._clientId}:${this._clientSecret}`).toString(
+          'base64'
+        );
+      headers['Authorization'] = basic_auth;
+    } else if (
+      this.client_authentication === ClientAuthentication.ClientSecretPost
+    ) {
+      Object.assign(values, {client_secret: this._clientSecret});
```

Comment:
```suggestion
      values.client_secret = this._clientSecret;
```

Reply:
>  The compiler gives me error of `Property 'client_secret' does not exist on type 'Omit<GetTokenOptions, "codeVerifier">'.ts(2339)`

Reply:
We can add the type to `values`:

```
values: Omit<GetTokenOptions, 'codeVerifier'> & {client_secret: string}
```

Essentially, we don't want properties that are difficult to track down in the event of an error or general debugging.